packages/perpetual/src/fundingRateProposer.js

```diff
@@ -0,0 +1,223 @@
+        //                                      within this error % will be requested to be updated,
+        //                                      and proposed to.
+        //                                      e.g. 0.05 implies 5% margin of error.
+        value: 0.05,
+        isValid: x => {
+          return x >= 0 && x < 1;
+        }
+      }
+    };
+
+    // Validate and set config settings to class state.
+    const configWithDefaults = createObjectFromDefaultProps(perpetualProposerConfig, defaultConfig);
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
+    await Promise.all([this.perpetualFactoryClient.update(), this.gasEstimator.update()]);
+
+    // Create contract object for each perpetual address created.
+    for (let creationEvent of this.perpetualFactoryClient.getAllCreatedContractEvents()) {
+      if (!this.contractCache[creationEvent.contractAddress]) {
+        // Failure to construct a Perpetual instance using the contract address should be fatal,
+        // so we don't catch that error.
+        const perpetualContract = this.createPerpetualContract(creationEvent.contractAddress);
+        this.contractCache[creationEvent.contractAddress] = perpetualContract;
+      }
+    }
+
+    // Increase allowances for all relevant collateral currencies.
+    await this._setAllowances();
+  }
+
+  async updateFundingRates() {
+    this.logger.debug({
+      at: "PerpetualProposer#updateFundingRates",
+      message: "Checking for contract funding rates to update"
+    });
+
+    // TODO: Should allow user to filter out price requests with rewards below a threshold,
+    // allowing the bot to prevent itself from being induced to unprofitably propose.
+    for (let contractAddress of Object.keys(this.contractCache)) {
+      await this._updateFundingRate(contractAddress);
+    }
+  }
+
+  /** **********************************
+   *
+   * INTERNAL METHODS
+   *
+   ************************************/
+
+  // Check contract funding rates and request+propose to update them, or return early if an error is encountered.
+  async _updateFundingRate(contractAddress) {
+    // Fetch current funding rate data from contract
+    const currentFundingRateData = await this.contractCache[contractAddress].methods.fundingRate().call({
+      from: this.account,
+      gasPrice: this.gasEstimator.getCurrentFastPrice()
+    });
+    const fundingRateIdentifier = currentFundingRateData.identifier;
+    const priceFeed = await this._createOrGetCachedPriceFeed(fundingRateIdentifier);
+
+    // Pricefeed is either constructed correctly or is null.
+    if (!priceFeed) {
+      this.logger.error({
+        at: "PerpetualProposer#updateFundingRates",
+        message: "Failed to construct a PriceFeed for funding rate identifier",
+        fundingRateIdentifier
+      });
+      return;
+    }
+
+    // With pricefeed successfully constructed, get the current funding rate
+    await priceFeed.update();
```

Comment:
do we not want to group the updates into a `Promise.all` in the `update` method?

Reply:
see [this comment](https://github.com/UMAprotocol/protocol/pull/2632#discussion_r586364110)